Fixing Buffer::BlockCopy, JIT_MemCpy, and JIT_MemSet to just call the appropriate CRT functions for x64 Windows, as is already done for all other platforms/targets

tannergooding opened the pull request:
This resolves https://github.com/dotnet/coreclr/issues/25505.

Based on the original issue where `JIT_MemCpy` was changed to use `rep movsb` (see https://github.com/dotnet/coreclr/pull/7198), there was:
* minor improvement (~5%) for arrays of length 0 to 120
* good improvement (~40%) for arrays of length 130 to 510
  * ~47% for arrays of length 130 to 310
  * ~39% for arrays of length 320 to 440
  * ~27% for arrays of length 450 to 510
* little improvement (~1%) for arrays above 510 in length
  * This was only tested for 520 and 1000 bytes

However, on AMD processors, there are additional limitations around `rep movsb` and when it is beneficial to use. The common conditions under which it is being used in the `JIT_MemCpy` method today actually cause a 2x perf decrease for arrays larger than 512 bytes.

Having a custom memcpy routine adds additional maintenance burden, can be error prone, is generally not as widely tested, and does not get many of the optimizations that the CRT implementations receives. This, coupled with the overall minor improvements for small arrays on Intel processors and the 2x regression for arrays over 512 bytes on AMD processors is resulting in the custom memcpy routine being removed.

It would be beneficial for any future improvements to `memcpy` to be made directly against `glibc` and `crt` instead.

tannergooding commented:
I did a `git grep -in JIT_Mem` and it looks like there are no actual remaining usages of `JIT_MemSet` or `JIT_MemCpy` in the repo:
```
C:\repos\coreclr [fix-25505 ≡]> git grep -in JIT_Mem                                                                                                                        src/inc/jithelpers.h:260:    JITHELPER(CORINFO_HELP_MEMSET,              JIT_MemSet,         CORINFO_HELP_SIG_REG_ONLY)                                                     src/inc/jithelpers.h:261:    JITHELPER(CORINFO_HELP_MEMCPY,              JIT_MemCpy,         CORINFO_HELP_SIG_REG_ONLY)
src/vm/amd64/CrtHelpers.asm:43:LEAF_ENTRY JIT_MemSet, _TEXT
src/vm/amd64/CrtHelpers.asm:149:LEAF_END_MARKED JIT_MemSet, _TEXT
src/vm/amd64/CrtHelpers.asm:151:;JIT_MemCpy - Copy source buffer to destination buffer
src/vm/amd64/CrtHelpers.asm:154:;   JIT_MemCpy() copies a source memory buffer to a destination memory
src/vm/amd64/CrtHelpers.asm:183:LEAF_ENTRY JIT_MemCpy, _TEXT
src/vm/amd64/CrtHelpers.asm:329:LEAF_END_MARKED JIT_MemCpy, _TEXT
src/vm/amd64/crthelpers.S:9:// JIT_MemSet/JIT_MemCpy
src/vm/amd64/crthelpers.S:16:LEAF_ENTRY JIT_MemSet, _TEXT
src/vm/amd64/crthelpers.S:27:LEAF_END_MARKED JIT_MemSet, _TEXT
src/vm/amd64/crthelpers.S:29:LEAF_ENTRY JIT_MemCpy, _TEXT
src/vm/amd64/crthelpers.S:41:LEAF_END_MARKED JIT_MemCpy, _TEXT
src/vm/arm/CrtHelpers.asm:23:; JIT_MemSet/JIT_MemCpy
src/vm/arm/CrtHelpers.asm:31:;EXTERN_C void __stdcall JIT_MemSet(void* _dest, int c, size_t count)
src/vm/arm/CrtHelpers.asm:32:        LEAF_ENTRY JIT_MemSet
src/vm/arm/CrtHelpers.asm:101:        LEAF_END_MARKED JIT_MemSet
src/vm/arm/CrtHelpers.asm:104:;EXTERN_C void __stdcall JIT_MemCpy(void* _dest, const void *_src, size_t count)
src/vm/arm/CrtHelpers.asm:105:        LEAF_ENTRY JIT_MemCpy
src/vm/arm/CrtHelpers.asm:159:        LEAF_END_MARKED JIT_MemCpy
src/vm/arm/crthelpers.S:21:// JIT_MemSet/JIT_MemCpy
src/vm/arm/crthelpers.S:29://EXTERN_C void __stdcall JIT_MemSet(void* _dest, int c, size_t count)
src/vm/arm/crthelpers.S:30:LEAF_ENTRY JIT_MemSet, _TEXT
src/vm/arm/crthelpers.S:40:LEAF_END_MARKED JIT_MemSet, _TEXT
src/vm/arm/crthelpers.S:43://EXTERN_C void __stdcall JIT_MemCpy(void* _dest, const void *_src, size_t count)
src/vm/arm/crthelpers.S:44:LEAF_ENTRY JIT_MemCpy, _TEXT
src/vm/arm/crthelpers.S:56:LEAF_END_MARKED JIT_MemCpy, _TEXT
src/vm/arm64/crthelpers.S:7:// JIT_MemSet/JIT_MemCpy
src/vm/arm64/crthelpers.S:13:LEAF_ENTRY JIT_MemSet, _TEXT
src/vm/arm64/crthelpers.S:14:    cbz x2, LOCAL_LABEL(JIT_MemSet_ret)
src/vm/arm64/crthelpers.S:20:LOCAL_LABEL(JIT_MemSet_ret):
src/vm/arm64/crthelpers.S:22:LEAF_END_MARKED JIT_MemSet, _TEXT
src/vm/arm64/crthelpers.S:24:LEAF_ENTRY JIT_MemCpy, _TEXT
src/vm/arm64/crthelpers.S:25:    cbz x2, LOCAL_LABEL(JIT_MemCpy_ret)
src/vm/arm64/crthelpers.S:32:LOCAL_LABEL(JIT_MemCpy_ret):
src/vm/arm64/crthelpers.S:34:LEAF_END_MARKED JIT_MemCpy, _TEXT
src/vm/arm64/crthelpers.asm:15:;void JIT_MemSet(void *dst, int val, SIZE_T count)
src/vm/arm64/crthelpers.asm:58:; Assembly code corresponding to above C++ method. JIT_MemSet can AV and clr exception personality routine needs to
src/vm/arm64/crthelpers.asm:59:; determine if the exception has taken place inside JIT_Memset in order to throw corresponding managed exception.
src/vm/arm64/crthelpers.asm:60:; Determining this is slow if the method were implemented as C++ method (using unwind info). In .asm file by adding JIT_MemSet_End
src/vm/arm64/crthelpers.asm:61:; marker it can be easily determined if exception happened in JIT_MemSet. Therefore, JIT_MemSet has been written in assembly instead of
src/vm/arm64/crthelpers.asm:64:    LEAF_ENTRY JIT_MemSet
src/vm/arm64/crthelpers.asm:70:    b           JIT_MemSet_bottom
src/vm/arm64/crthelpers.asm:71:JIT_MemSet_top
src/vm/arm64/crthelpers.asm:73:JIT_MemSet_bottom
src/vm/arm64/crthelpers.asm:75:    bge        JIT_MemSet_top
src/vm/arm64/crthelpers.asm:77:    tbz         x2, #3, JIT_MemSet_tbz4
src/vm/arm64/crthelpers.asm:79:JIT_MemSet_tbz4
src/vm/arm64/crthelpers.asm:80:    tbz         x2, #2, JIT_MemSet_tbz2
src/vm/arm64/crthelpers.asm:82:JIT_MemSet_tbz2
src/vm/arm64/crthelpers.asm:83:    tbz         x2, #1, JIT_MemSet_tbz1
src/vm/arm64/crthelpers.asm:85:JIT_MemSet_tbz1
src/vm/arm64/crthelpers.asm:86:    tbz         x2, #0, JIT_MemSet_ret
src/vm/arm64/crthelpers.asm:88:JIT_MemSet_ret
src/vm/arm64/crthelpers.asm:92:    LEAF_ENTRY JIT_MemSet_End
src/vm/arm64/crthelpers.asm:97:; See comments above for JIT_MemSet
src/vm/arm64/crthelpers.asm:99:;void JIT_MemCpy(void *dst, const void *src, SIZE_T count)
src/vm/arm64/crthelpers.asm:143:; See comments above for JIT_MemSet method
src/vm/arm64/crthelpers.asm:144:    LEAF_ENTRY JIT_MemCpy
src/vm/arm64/crthelpers.asm:145:    b           JIT_MemCpy_bottom
src/vm/arm64/crthelpers.asm:146:JIT_MemCpy_top
src/vm/arm64/crthelpers.asm:149:JIT_MemCpy_bottom
src/vm/arm64/crthelpers.asm:151:    bge         JIT_MemCpy_top
src/vm/arm64/crthelpers.asm:153:    tbz         x2, #3, JIT_MemCpy_tbz4
src/vm/arm64/crthelpers.asm:156:JIT_MemCpy_tbz4
src/vm/arm64/crthelpers.asm:157:    tbz         x2, #2, JIT_MemCpy_tbz2
src/vm/arm64/crthelpers.asm:160:JIT_MemCpy_tbz2
src/vm/arm64/crthelpers.asm:161:    tbz         x2, #1, JIT_MemCpy_tbz1
src/vm/arm64/crthelpers.asm:164:JIT_MemCpy_tbz1
src/vm/arm64/crthelpers.asm:165:    tbz         x2, #0, JIT_MemCpy_ret
src/vm/arm64/crthelpers.asm:168:JIT_MemCpy_ret
src/vm/arm64/crthelpers.asm:172:    LEAF_ENTRY JIT_MemCpy_End
src/vm/excep.cpp:6702:EXTERN_C void JIT_MemSet_End();
src/vm/excep.cpp:6703:EXTERN_C void JIT_MemCpy_End();
src/vm/excep.cpp:6762:    CHECK_RANGE(JIT_MemSet)
src/vm/excep.cpp:6763:    CHECK_RANGE(JIT_MemCpy)
src/vm/exceptionhandling.cpp:5213:        // in one of the JIT helper functions (JIT_MemSet, ...)
src/vm/jitinterface.h:434:    void STDCALL JIT_MemSet(void *dest, int c, SIZE_T count);
src/vm/jitinterface.h:435:    void STDCALL JIT_MemCpy(void *dest, const void *src, SIZE_T count);
```

Should we consider just removing the assembly helpers and the `CHECK_RANGE` sanity checks that exist?

tannergooding commented:
CC. @jkotas, @janvorli

janvorli commented:
JIT still uses these helpers, see:
https://github.com/dotnet/coreclr/blob/dbc0e19e87e1ac6e2be33a1ddeea95f21ba2bc39/src/inc/jithelpers.h#L260-L261

The CHECK_RANGE are not sanity checks, they are actually necessary. In case the JIT_MemCpy or JIT_MemSet get null reference as an argument, we need to handle that as if it happened in the managed caller.

janvorli commented:
To use memcpy / memset from the CRT, we should do the same thing that we do on AMD64 Linux:
https://github.com/dotnet/coreclr/blob/dbc0e19e87e1ac6e2be33a1ddeea95f21ba2bc39/src/vm/amd64/crthelpers.S#L16-L39
That way, we check the input parameters in the JIT_xxx functions and then tail call to the runtime functions.

tannergooding commented:
Fixed up the JIT_MemSet and JIT_MemCpy implementations to forward to the CRT versions. I also cleaned up the comments in the Unix version to match those in the x64 version (where applicable).

tannergooding commented on src/vm/amd64/crthelpers.S:
```diff
@@ -13,27 +13,61 @@
+        cmp     byte ptr [rsi], 0       // check src for null
 
-        jmp  C_PLTFUNC(memcpy)
+        jmp     C_PLTFUNC(memmove)      // forward to the CRT implementation, using memmove to handle overlapping buffers
```
This was the only code change to the Unix version. The Windows version was very explicit about needing to handle overlapping buffers, and so `memmove` is the appropriate thing to forward to.

I would assume the same conditions apply to Unix.

jkotas replied:
> The Windows version was very explicit about needing to handle overlapping buffers

That was for better compatibility with .NET Framework. It is not required by the ECMA 335: The behavior of cpblk is unspecified if the source and destination areas overlap.`

memmove is a bit slower. I do not think we need to be making the JIT helper slower on Unix...

tannergooding replied:
Do we still need it for compat on Windows? I would think consistency here is also important.

(I'll keep `memmove` on Windows and `memcpy` on Unix for now, but I think it would  be a good discussion to have).

tannergooding replied:
If nothing else, it might be good to have `JIT_MemCpy` and `JIT_MemMove`, and then the JIT can more readily decide which is appropriate based on the scenario. There may be scenarios where overlapping is desirable and several where it is not needed. And, it appears as though more than just `cpblk` goes through these.

jkotas replied:
IIRC, memcpy as alias for memmove in Windows x64 CRT. So even if we have changed Windows to call memcpy, the unspecified behavior is still going to be different between Windows and Unix.

tannergooding commented on src/vm/amd64/CrtHelpers.asm:
```diff
@@ -14,317 +14,74 @@
-;   void *dst = pointer to destination buffer
-;   const void *src = pointer to source buffer
-;   size_t count = number of bytes to copy
+; Exit:
```
The declaration of these methods in `jitinterface.h` have them as void returning, rather than as `void*` returning (which is what the CRT implementations do):
https://github.com/dotnet/coreclr/blob/01da3bdd734c887856f89037cd4a3f2b5a072967/src/vm/jitinterface.h#L434

tannergooding commented:
CoreFX legs are failing due to:
```
It was not possible to find any compatible framework version
The specified framework 'Microsoft.NETCore.App', version '3.0.0' was not found.
  - The following frameworks were found:
      5.0.0 at [/home/helixbot/work/8698dcd6-71bd-4dec-90db-5aaae809ddac/Payload/shared/Microsoft.NETCore.App]
```

@wtgodbe, was this something missed or are things just not entirely in sync yet?

wtgodbe commented:
We may need to update some TFMs to `netcoreapp5.0` - see https://github.com/dotnet/coreclr/pull/25704#issuecomment-511933013. @BruceForstall where are the `runtimeconfig.json` files you mentioned generated? @ericstj, is it alright if we hard-code a 5.0 TFM?

BruceForstall commented:
@wtgodbe The runtimeconfig.json files are built with the corefx tests in the corefx official build, and downloaded by the corefx test legs in coreclr repo CI.

Maybe the 3.0 versions here: https://github.com/dotnet/coreclr/blob/master/tests/src/Common/CoreFX/CoreFX.depproj need to be updated?

tannergooding commented:
Tests aren't running...

tannergooding commented:
/azp help

tannergooding commented:
Closed and reopened as per https://github.com/dotnet/corefx/issues/39593
